tests/input.rs

```diff
@@ -0,0 +1,299 @@
+use bevy::{
+    asset::AssetPlugin,
+    core::CorePlugin,
+    core_pipeline::CorePipelinePlugin,
+    ecs::event::Events,
+    input::InputPlugin,
+    input::{keyboard::KeyboardInput, ButtonState},
+    pbr::PbrPlugin,
+    prelude::{App, Entity},
+    prelude::{Commands, Component, EventReader, KeyCode},
+    render::RenderPlugin,
+    utils::HashSet,
+    window::{ReceivedCharacter, WindowId, WindowPlugin},
+};
+
+#[derive(Component)]
+struct CharComponent(char);
+fn spawn_entity_at_char_event_system(
+    mut commands: Commands,
+    mut char_input_events: EventReader<ReceivedCharacter>,
+) {
+    for event in char_input_events.iter() {
+        commands.spawn().insert(CharComponent(event.char));
+    }
+}
+
+#[derive(Component)]
+struct KeyComponent(KeyboardInput);
+fn spawn_entity_at_keyboard_event_system(
+    mut commands: Commands,
+    mut char_input_events: EventReader<KeyboardInput>,
+) {
+    for event in char_input_events.iter() {
+        commands.spawn().insert(KeyComponent(event.clone()));
+    }
+}
+
+macro_rules! get_entities {
+    ($app: ident, $query: tt) => {{
+        let world = &mut $app.world;
+        let mut query = world.query::<$query>();
+        query.iter(world).collect::<Vec<_>>()
+    }};
+}
+
+macro_rules! send_event {
+    ($app: ident, $t: ty, $event: expr) => {
+        let world = &mut $app.world;
+        let mut sender = world.get_resource_mut::<Events<$t>>().unwrap();
+        sender.send($event);
+    };
+}
+
+#[cfg_attr(target_os = "linux")]
+#[test]
+fn test_input_received_character_single_button() {
+    let mut app = App::new();
+
+    app.add_plugin(CorePlugin::default());
+    app.add_plugin(WindowPlugin::default());
+    app.add_plugin(AssetPlugin::default());
+    app.add_plugin(RenderPlugin::default());
+    app.add_plugin(CorePipelinePlugin::default());
+    app.add_plugin(PbrPlugin::default());
+
+    app.add_system(spawn_entity_at_char_event_system);
+
+    app.update();
+
+    let entities: Vec<_> = get_entities!(app, Entity);
+    assert_eq!(entities.len(), 0);
+
+    send_event!(
+        app,
+        ReceivedCharacter,
+        ReceivedCharacter {
+            id: WindowId::primary(),
+            char: 'a',
+        }
+    );
```

Comment:
I don't think this macro is needed

Reply:
the macros simplify the readability of the test.
Avoids also the copy-and-paste.
Should I remove it ?

Reply:
I would prefer to avoid macros here.